### ISSUE TITLE
Clean up memory in Drake workflow R

### ISSUE THREAD

**Issue by edgBR** [ropensci/drake#1293]
## Description

Following the advice of Will, I am opening this new issue as we discussed here: https://stackoverflow.com/questions/62731127/clean-up-memory-in-drake-workflow-r/62732757?noredirect=1#comment110953104_62732757

## Reproducible example

I have package a massive time series workflow (4273*10 models) for 4273 time series weekly in drake.

Originally I attempted to create the full workflow using the fable package. Which is quite handy to train models for grouped tsibbles, but after different trials I got many many problems with memory management. My RStudio server with 32 cores and 244 GBs of RAM was crashing constantly specially when I was trying to serialize the models.

Because of that I completely spitted my workflow in order to identify bottlenecks going from:

[![enter image description here][1]][1]

To:

[![enter image description here][2]][2]

Then to:

[![enter image description here][3]][3]

An finally to:

[![enter image description here][4]][4]

Inside my training code (example prophet_multiplicative) I am using the future package to train this multiple fable models and then calculate the accuracy and save them. However I am not aware how to remove this object from the drake workflow afterwards:

 - Should I just remove the object using rm?  
 - Is there any way in drake to have separate environments for each of the workflow components? 
 - Is this the right solution?

My idea is to run each of individual techniques in a serial manner meanwhile the 4273 models for one specific technique are trained in parallel. Doing so I expect to not crash the server and then after all of my models are trained I can join the accuracy metrics, pick the best model for each of my time series and then trim each of the individual binary files to be able to produce the forecasts.

  [1]: https://i.stack.imgur.com/V70ju.png
  [2]: https://i.stack.imgur.com/6CRD2.png
  [3]: https://i.stack.imgur.com/kBnzt.png
  [4]: https://i.stack.imgur.com/mgnJN.png

**Start from HERE:**

After the first comments from Will my make plan now looks like this (I have add a remote cache because my main disk is full):

``` R
remote_cache <- new_cache(path = paste0("/data1/.drake")) # File path to drake's cache.


make(plan = plan,
     seed = 666,
     cache = remote_cache,
     verbose = 2, 
     log_progress = TRUE,
     recover = TRUE,
     lock_envir = FALSE,
     garbage_collection = TRUE,
     memory_strategy = "autoclean")

```

And my plan is as follows:

```R
plan <- drake_plan(
  data = getLifeData(environment = "PROD", 
                                         key_directory = config_parameters$LOCAL_CONFIG$DirectoryKeyCloud_RStudio, 
                                         max_forecasting_horizon = argument_parser$horizon),
  metadata = getMetadata(environment = "PROD", 
                              key_directory = config_parameters$LOCAL_CONFIG$DirectoryKeyCloud_RStudio, 
                              schema = config_parameters$SF_CONFIG$schema_name, db_src = c(1,2,3)),
  recent_data = getLastData(data),
  processed_data = featureEngineering(raw_data = data, 
                                      metadata = metadata, 
                                      recent = recent_data,
                                      max_forecasting_horizon = argument_parser$horizon),
  prophet_multiplicative = trainModels(input_data = processed_data, 
                               max_forecast_horizon = argument_parser$horizon,
                               max_multisession_cores = argument_parser$sessions,
                               model_type = "prophet_multiplicative"),
  accuracy_prophet_multiplicative = accuracy_explorer(mode = "train", models = prophet_multiplicative, 
                                                      max_forecast_horizon = argument_parser$horizon,
                                                      directory_out = "/data1/temp_rds_storage/"),
  saving_prophet_multiplicative = saveModels(models = prophet_multiplicative, 
                       directory_out = "/data1/temp_rds_storage/",
                       max_forecasting_horizon = argument_parser$horizon,
                       max_multisession_cores = argument_parser$sessions),
  prophet_additive = trainModels(input_data = processed_data, 
                                       max_forecast_horizon = argument_parser$horizon,
                                       max_multisession_cores = argument_parser$sessions,
                                       model_type = "prophet_additive"),
  accuracy_prophet_additive = accuracy_explorer(mode = "train", models = prophet_additive, 
                                                      max_forecast_horizon = argument_parser$horizon,
                                                      directory_out = "/data1/temp_rds_storage/"),
  saving_prophet_additive = saveModels (models = prophet_additive, 
                                              directory_out = "/data1/temp_rds_storage/",
                                              max_forecasting_horizon = argument_parser$horizon,
                                              max_multisession_cores = argument_parser$sessions),
# AUTO-ARIMA fails for some unexplinable reason
  # auto_arima = trainModels(input_data = processed_data, 
  #                                max_forecast_horizon = argument_parser$horizon,
  #                                max_multisession_cores = argument_parser$sessions,
  #                                model_type = "auto_arima"),
  # accuracy_auto_arima = accuracy_explorer(mode = "train", 
  #                                         models = auto_arima, 
  #                                         max_forecast_horizon = argument_parser$horizon, 
  #                                         directory_out = "/data1/temp_rds_storage/"),
  # saving_auto_arima = saveModels (models = auto_arima, 
  #                                       directory_out = "/data1/temp_rds_storage/",
  #                                       max_forecasting_horizon = argument_parser$horizon,
  #                                       max_multisession_cores = argument_parser$sessions),
  auto_arima_yearly_fourier = trainModels(input_data = processed_data, 
                           max_forecast_horizon = argument_parser$horizon,
                           max_multisession_cores = argument_parser$sessions,
                           model_type = "arima_with_yearly_fourier_components"),
  accuracy_auto_arima_yearly_fourier = accuracy_explorer(mode = "train", models = auto_arima_yearly_fourier, 
                                                max_forecast_horizon = argument_parser$horizon,
                                                directory_out = "/data1/temp_rds_storage/"),
  saving_auto_arima_yearly_fourier = saveModels (models = auto_arima_yearly_fourier, 
                                  directory_out = "/data1/temp_rds_storage/",
                                  max_forecasting_horizon = argument_parser$horizon,
                                  max_multisession_cores = argument_parser$sessions),
  auto_arima_monthly_fourier = trainModels(input_data = processed_data, 
                                           max_forecast_horizon = argument_parser$horizon, 
                                           max_multisession_cores = argument_parser$sessions,
                                           model_type = "arima_with_monthly_fourier_components"),
  accuracy_auto_arima_monthly_fourier = accuracy_explorer(mode = "train", models = auto_arima_monthly_fourier, 
                                                         max_forecast_horizon = argument_parser$horizon,
                                                         directory_out = "/data1/temp_rds_storage/"),
  saving_auto_arima_monthly_fourier = saveModels (models = auto_arima_monthly_fourier, 
                                                  directory_out = "/data1/temp_rds_storage/",
                                                  max_forecasting_horizon = argument_parser$horizon,
                                                  max_multisession_cores = argument_parser$sessions),
  reg_arima_errors = trainModels(input_data = processed_data, 
                                max_forecast_horizon = argument_parser$horizon,
                                max_multisession_cores = argument_parser$sessions,
                                model_type = "regression_with_arima_errors"),
  accuracy_reg_arima_errors = accuracy_explorer(mode = "train", 
                                                models = reg_arima_errors, 
                                                max_forecast_horizon = argument_parser$horizon,
                                                directory_out = "/data1/temp_rds_storage/"),
  saving_reg_arima_errors = saveModels (models = reg_arima_errors, 
                                        directory_out = "/data1/temp_rds_storage/", 
                                        max_forecasting_horizon = argument_parser$horizon, 
                                        max_multisession_cores = argument_parser$sessions),
  tslm_model = trainModels(input_data = processed_data, 
                           max_forecast_horizon = argument_parser$horizon, 
                           max_multisession_cores = argument_parser$sessions,
                          model_type = "tslm"),
  accuracy_tslm = accuracy_explorer(mode = "train", 
                                    models = tslm_model, 
                                    max_forecast_horizon = argument_parser$horizon,
                                    directory_out = "/data1/temp_rds_storage/"),
  saving_tslm = saveModels (models = tslm_model, 
                            directory_out = "/data1/temp_rds_storage/", 
                            max_forecasting_horizon = argument_parser$horizon, 
                            max_multisession_cores = argument_parser$sessions),
  # theta_model = trainModels(input_data = processed_data, 
  #                               max_forecast_horizon = argument_parser$horizon,
  #                               max_multisession_cores = argument_parser$sessions,
  #                               model_type = "theta"),
  # accuracy_theta = accuracy_explorer(type = "train", 
  #                                    models = theta_model, 
  #                                    max_forecast_horizon = argument_parser$horizon,
  #                                    directory_out = "/data1/temp_rds_storage/"),
  # saving_theta = saveModels (models = theta_model, 
  #                                       directory_out = "/data1/temp_rds_storage/", 
  #                                       max_forecasting_horizon = argument_parser$horizon, 
  #                                       max_multisession_cores = argument_parser$sessions)
  
)
```

All of the functions work perfectly outside drake but for some reason auto_arima and theta are not working and I am getting the following error:

```
[1] "DB PROD Connected"
[1] "DB PROD Connected"
[1] "Getting RAW data"
[1] "Maximum forecasting horizon is 52, fetching weekly data"
[1] "Removing duplicates if we have them"
[1] "Original data has 1860590 rows"
[1] "Data without duplicates has 1837995 rows"
`summarise()` regrouping output by 'A', 'B' (override with `.groups` argument)
[1] "Removing non active customers"
[1] "Data without duplicates and without active customers has 1654483 rows"
0.398 sec elapsed
[1] "Removing customers with last data older than 1.5 years"
[1] "Data without duplicates, customers that are not active and old customers has 1268610 rows"
0.223 sec elapsed
[1] "Augmenting data"
12.103 sec elapsed
[1] "Creating tsibble"
7.185 sec elapsed
[1] "Filling gaps for not breaking groups"
9.568 sec elapsed
[1] "Training theta models for forecasting horizon 52"
[1] "Using 12 sessions from as future::plan()"
Repacking large object
[1] "Training auto_arima models for forecasting horizon 52"
[1] "Using 12 sessions from as future::plan()"
Error: target auto_arima failed.
diagnose(auto_arima)error$message:
  object 'ts_models' not found
diagnose(auto_arima)error$calls:
  1. └─global::trainModels(...)
In addition: Warning message:
9 errors (2 unique) encountered for theta
[3] function cannot be evaluated at initial parameters
[6] Not enough data to estimate this ETS model.

Execution halted
```
My training function looks as follows:

```R
trainModels <- function(input_data, max_forecast_horizon, model_type, max_multisession_cores) {

  options(future.globals.maxSize = 1500000000)
  future::plan(multisession, workers = max_multisession_cores) #breaking infrastructure once again ;)
  set.seed(666) # reproducibility
  
    if(max_forecast_horizon <= 104) {
      
      print(paste0("Training ", model_type, " models for forecasting horizon ", max_forecast_horizon))
      print(paste0("Using ", max_multisession_cores, " sessions from as future::plan()"))
      
      if(model_type == "prophet_multiplicative") {
        
        ts_models <- input_data %>% model(prophet = fable.prophet::prophet(snsr_val_clean ~ season("week", 2, type = "multiplicative") + 
                                                                             season("month", 2, type = "multiplicative")))
        
      } else if(model_type == "prophet_additive") {
        
        ts_models <- input_data %>% model(prophet = fable.prophet::prophet(snsr_val_clean ~ season("week", 2, type = "additive") + 
                                                                             season("month", 2, type = "additive")))
        
      } else if(model_type == "auto.arima") {
        
        ts_models <- input_data %>% model(auto_arima = ARIMA(snsr_val_clean))
        
      } else if(model_type == "arima_with_yearly_fourier_components") {
        
        ts_models <- input_data %>% model(auto_arima_yf = ARIMA(snsr_val_clean ~ fourier("year", K = 2)))
        
      } else if(model_type == "arima_with_monthly_fourier_components") {
        
        ts_models <- input_data %>% model(auto_arima_mf = ARIMA(snsr_val_clean ~ fourier("month", K=2)))
        
      } else if(model_type == "regression_with_arima_errors") {
        
        ts_models <- input_data %>% model(auto_arima_mf_reg = ARIMA(snsr_val_clean ~ month + year  + quarter + qday + yday + week))
        
      } else if(model_type == "tslm") {
    
        ts_models <- input_data %>% model(tslm_reg_all = TSLM(snsr_val_clean ~ year  + quarter + month + day + qday + yday + week + trend()))
     
      } else if(model_type == "theta") {
        
        ts_models <- input_data %>% model(theta = THETA(snsr_val_clean ~ season()))
        
      } else if(model_type == "ensemble") {
        
        ts_models <- input_data %>% model(ensemble =  combination_model(ARIMA(snsr_val_clean), 
                                              ARIMA(snsr_val_clean ~ fourier("month", K=2)),
                                              fable.prophet::prophet(snsr_val_clean ~ season("week", 2, type = "multiplicative") +
                                              season("month", 2, type = "multiplicative"), 
                                              theta = THETA(snsr_val_clean ~ season()), 
                                              tslm_reg_all = TSLM(snsr_val_clean ~ year  + quarter + month + day + qday + yday + week + trend())))
            )
        
      }
      
    } 
  
    else if(max_forecast_horizon > 104) {
      
       ### STUFF FOR MONTHLY FORECAST
        }
    }
  
  return(ts_models)
}

```

Any clue of why drake is not fetching the processed data in the training function of auto_arima?

BR
/E

**Comment by wlandau** [comment]
Thanks for posting the, having it definitely helps.

One thing I notice right away is that your plan sets `model_type = "auto_arima"` in `trainModels()`,  but the `trainModels()` function itself checks for the string `"auto.arima"` instead. So could just be just a matter of changing a dot to an underscore in the function.


```r
auto_arima = trainModels(
    input_data = processed_data, 
    max_forecast_horizon = argument_parser$horizon,
    max_multisession_cores = argument_parser$sessions,
    model_type = "auto_arima"
  )

...

 } else if(model_type == "auto.arima") {
        
        ts_models <- input_data %>% model(auto_arima = ARIMA(snsr_val_clean))
        
      } 
```


If that doesn't explain it, let's try to take what you have posted and work it into a reproducible example. A good reproducible example should make it as easy as possible for the package maintainer to understand your code and recreate the error. So let's try to rework what you posted into a form that I can understand and run.

The following smaller plan should be enough to recreate the error, correct?

```{r}
plan <- drake_plan(
  data = getLifeData(
    environment = "PROD", 
    key_directory = config_parameters$LOCAL_CONFIG$DirectoryKeyCloud_RStudio, 
    max_forecasting_horizon = argument_parser$horizon)
  ,
  metadata = getMetadata(
    environment = "PROD", 
    key_directory = config_parameters$LOCAL_CONFIG$DirectoryKeyCloud_RStudio, 
    schema = config_parameters$SF_CONFIG$schema_name, db_src = c(1,2,3)
  ),
  recent_data = getLastData(data),
  processed_data = featureEngineering(
    raw_data = data, 
    metadata = metadata, 
    recent = recent_data,
    max_forecasting_horizon = argument_parser$horizon
  ),
  auto_arima = trainModels(
    input_data = processed_data, 
    max_forecast_horizon = argument_parser$horizon,
    max_multisession_cores = argument_parser$sessions,
    model_type = "auto_arima"
  )
)
```

Next, we can simplify `trainModels()` by just using the auto arima part and removing `future`.

```{r}
trainModels <- function(input_data, max_forecast_horizon, model_type, max_multisession_cores) {
  set.seed(666) # reproducibility
  input_data %>%
    model(auto_arima = ARIMA(snsr_val_clean))
}
```

Next, I need the other functions and packages referenced in the plan and the `trainModels()` function so I can try to reproduce what you see. With the information I have so far, the pipeline errors out because some functions have not been defined yet.

```{r}
make(plan)
▶ target metadata
x fail metadata
Error: target metadata failed.
diagnose(metadata)$error$message:
  could not find function "getMetadata"
diagnose(metadata)$error$calls:
```

**Comment by wlandau** [comment]
For more on reproducible examples, [this post](https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example) has some helpful advice and links.

**Comment by edgBR** [comment]
Hi @wlandau 

Now is working!

Thanks for the support it was exactly the typo error. Before closing the issue I would like to know if you think that my current plan can be simplified with dynamic or static branching.

I have take a look to the docs but I have not come up with an implementation that works.

BR
/E

**Comment by wlandau** [comment]
Glad to hear it worked!

Since you have a relatively small number of models and informative target names are preferred, I think static branching could help here. Sketch:

``` r
library(drake)

types <- c(
  "arima",
  "tslm",
  "theta"
)

plan <- drake_plan(
  data = getLifeData(
    environment = "PROD", 
    key_directory = config_parameters$LOCAL_CONFIG$DirectoryKeyCloud_RStudio, 
    max_forecasting_horizon = argument_parser$horizon
  ),
  metadata = getMetadata(
    environment = "PROD", 
    key_directory = config_parameters$LOCAL_CONFIG$DirectoryKeyCloud_RStudio, 
    schema = config_parameters$SF_CONFIG$schema_name, db_src = c(1,2,3)
  ),
  recent_data = getLastData(data),
  processed_data = featureEngineering(
    raw_data = data, 
    metadata = metadata, 
    recent = recent_data,
    max_forecasting_horizon = argument_parser$horizon
  ),
  model = target(
    trainModels(
      input_data = processed_data, 
      max_forecast_horizon = argument_parser$horizon,
      max_multisession_cores = argument_parser$sessions,
      model_type = type
    ),
    transform = map(type = !!types)
  ),
  accuracy = target(
    accuracy_explorer(
      type = "train", 
      models = model, 
      max_forecast_horizon = argument_parser$horizon,
      directory_out = "/data1/temp_rds_storage/"
    ),
    transform = map(model, .id = type)
  ),
  saving = target(
    saveModels(
      models = model, 
      directory_out = "/data1/temp_rds_storage/", 
      max_forecasting_horizon = argument_parser$horizon, 
      max_multisession_cores = argument_parser$sessions
    ),
    transform = map(model, .id = type)
  )
)

drake_plan_source(plan)
#> drake_plan(
#>   accuracy_arima = accuracy_explorer(
#>     type = "train", models = model_arima, max_forecast_horizon = argument_parser$horizon,
#>     directory_out = "/data1/temp_rds_storage/"
#>   ),
#>   accuracy_tslm = accuracy_explorer(
#>     type = "train", models = model_tslm, max_forecast_horizon = argument_parser$horizon,
#>     directory_out = "/data1/temp_rds_storage/"
#>   ),
#>   accuracy_theta = accuracy_explorer(
#>     type = "train", models = model_theta, max_forecast_horizon = argument_parser$horizon,
#>     directory_out = "/data1/temp_rds_storage/"
#>   ),
#>   data = getLifeData(
#>     environment = "PROD", key_directory = config_parameters$LOCAL_CONFIG$DirectoryKeyCloud_RStudio,
#>     max_forecasting_horizon = argument_parser$horizon
#>   ),
#>   metadata = getMetadata(
#>     environment = "PROD", key_directory = config_parameters$LOCAL_CONFIG$DirectoryKeyCloud_RStudio,
#>     schema = config_parameters$SF_CONFIG$schema_name, db_src = c(
#>       1,
#>       2, 3
#>     )
#>   ),
#>   model_arima = trainModels(
#>     input_data = processed_data, max_forecast_horizon = argument_parser$horizon,
#>     max_multisession_cores = argument_parser$sessions, model_type = "arima"
#>   ),
#>   model_tslm = trainModels(
#>     input_data = processed_data, max_forecast_horizon = argument_parser$horizon,
#>     max_multisession_cores = argument_parser$sessions, model_type = "tslm"
#>   ),
#>   model_theta = trainModels(
#>     input_data = processed_data, max_forecast_horizon = argument_parser$horizon,
#>     max_multisession_cores = argument_parser$sessions, model_type = "theta"
#>   ),
#>   processed_data = featureEngineering(
#>     raw_data = data, metadata = metadata, recent = recent_data,
#>     max_forecasting_horizon = argument_parser$horizon
#>   ),
#>   recent_data = getLastData(data),
#>   saving_arima = saveModels(
#>     models = model_arima, directory_out = "/data1/temp_rds_storage/",
#>     max_forecasting_horizon = argument_parser$horizon, max_multisession_cores = argument_parser$sessions
#>   ),
#>   saving_tslm = saveModels(
#>     models = model_tslm, directory_out = "/data1/temp_rds_storage/",
#>     max_forecasting_horizon = argument_parser$horizon, max_multisession_cores = argument_parser$sessions
#>   ),
#>   saving_theta = saveModels(
#>     models = model_theta, directory_out = "/data1/temp_rds_storage/",
#>     max_forecasting_horizon = argument_parser$horizon, max_multisession_cores = argument_parser$sessions
#>   )
#> )
```

<sup>Created on 2020-07-06 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

```{r}
plot(plan)
```

![Screen Shot 2020-07-06 at 2 03 16 PM](https://user-images.githubusercontent.com/1580860/86624544-79a15200-bf91-11ea-886f-bee5bc87f79b.png)
